### DISCUSSION TITLE
When an exception occurs when the lower-level pipeline is executed, how does the upper layer be notified?

### DISCUSSION THREAD

**Discussion by nicelulu** [databendlabs/databend#12041] [Q&A]
Hi, is it sent to the upper Exchange Source through the pipeline's own Exchange Sink, or is the exception notified directly to the coordinator, and the coordinator cancels the other pipelines and sends the exception to the user client?

**Comment by bohutang** [comment]
For this question, I would like to ping @JackTan25

**Comment by JackTan25** [comment]
@nicelulu Thanks for your question. I think the correlated code is here [exchange_manager](https://github.com/datafuselabs/databend/blob/main/src/query/service/src/api/rpc/exchange/exchange_manager.rs#L723). The path is below: 
    1. a node get a exception. it will pass it into the shutdown func. and then [send error](https://github.com/datafuselabs/databend/blob/main/src/query/service/src/api/rpc/exchange/statistics_sender.rs#L63)
    2. source node will receive the error here [get_error1](https://github.com/datafuselabs/databend/blob/main/src/query/service/src/api/rpc/exchange/statistics_receiver.rs#L79)
    or [get_error2](https://github.com/datafuselabs/databend/blob/main/src/query/service/src/api/rpc/exchange/statistics_receiver.rs#L98),it will notify the pipeline to finish. 
    So the answer of your  question is that the exchange source just receive data. All in all, all operators just receive data and process it and then output it. The error mechanism is controlled by higher level.

**Reply by nicelulu** [reply]
Thanks.
